hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterSnapshot.java

```diff
@@ -166,14 +167,30 @@ public void renameSnapshot(String snapshotRoot, String oldSnapshotName,
     RemoteMethod remoteMethod = new RemoteMethod("getSnapshotListing",
         new Class<?>[]{String.class},
         new RemoteParam());
+    SnapshotStatus[] response;
     if (rpcServer.isInvokeConcurrent(snapshotRoot)) {
       Map<RemoteLocation, SnapshotStatus[]> ret = rpcClient.invokeConcurrent(
           locations, remoteMethod, true, false, SnapshotStatus[].class);
-      return ret.values().iterator().next();
+      response = ret.values().iterator().next();
+      String src = ret.keySet().iterator().next().getSrc();
+      String dst = ret.keySet().iterator().next().getDest();
+      for (SnapshotStatus s : response) {
+        String mountPath =
+            new String(s.getParentFullPath()).replaceFirst(src, dst);
+        s.setParentFullPath(mountPath.getBytes());
+      }
     } else {
-      return rpcClient.invokeSequential(
+      response = rpcClient.invokeSequential(
           locations, remoteMethod, SnapshotStatus[].class, null);
+      RemoteLocation loc = locations.get(0);
```

Comment:
The actual location won't be always the first one in the list. `RouterRpcClient#L858` iterates over the list and triggers call to them sequentially, if the first location doesn't give the specific response, then second and so on. Adding a line at `RouterRpcClient` at `L872`, shall make this logic work -
``          Collections.swap(locations, 0, locations.indexOf(loc));``
Apart everything seems good here.

Reply:
Thanks @ayushtkn . I tried doing the similar change but it breaks many existing unit tests. 
Moreover, i see a similar path has been followed in createSnapshot as well. 

Do you think, createSnapshot/getSnapshottableDirectory Listing is broken as well?
 I am not very familiar with RouterProtocol code. In such a case, i would appreciate if you would fix/help fixing all of these at the same time in a separate jira?